### PR TITLE
Associating Pull Requests with Deploys and Showing Pull Requests

### PR DESCRIPTION
**Pull request by emmasax1**
Description and Impact
----------------------
This will allow users to see the pull requests on a deploys index view, as well as click to see more details about each pull request.

Deploy Plan
-----------
... I think migrations are present...

Rollback Plan
-------------
* To roll back this change, revert the merge with: `git revert -m 1 MERGE_SHA` and perform another deploy.

URLs
----
* [Jira issue 3056](https://sportngin.atlassian.net/secure/RapidBoard.jspa?rapidView=33&view=detail&selectedIssue=IS-3056)

QA Plan
-------
- [ ] Perhaps play with the new additions locally

- [ ] Go through code to ensure that it follows good styles and everything

### CODE REVIEW

**Comment by bbergstrom** [comment]
Please undo the massive find and replace on quotes you have done.  If you intend to do that, please make that part of a separate PR.

File: app/assets/javascripts/hubstats/deploys.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    updateQueryStringParameter(queryParameters,"repos",ids);
+  });
+  
+  $("#users").change(function() {
+    var ids = $("#users").val();
+    updateQueryStringParameter(queryParameters,"users",ids);
+  });
+});
+
+function updateQueryStringParameter(queryParameters, key, value) {
```
Comment (bbergstrom) [inline app/assets/javascripts/hubstats/deploys.js]: Are these function definitions duplicate of the ones in `pull_requests.js`?  If so, there is no need to duplicate them.  If they are different, you need to name them differently so they don't conflict.  The last one loaded will win the battle.

File: app/assets/javascripts/hubstats/deploys.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// Place all the behaviors and hooks related to the matching controller here.
+// All this logic will automatically be available in application.js.
+
+$(document).ready(function() {
```
Comment (bbergstrom) [inline app/assets/javascripts/hubstats/deploys.js]: Is this different from the ready function in pull_requests.js?

File: app/assets/stylesheets/hubstats/application.css [hunk 3]
```diff
@@ -82,7 +82,8 @@
   margin: 0px 0px 15px 0px;
 }
 
-.pulls > .col, .comments > .col, .users > .col, .repos > .col, .deploys > .col {
+.pulls > .col, .comments > .col, .users > .col, .repos > .col, .deploys > .col
```
Comment (bbergstrom) [inline app/assets/stylesheets/hubstats/application.css]: `.deploys > .col` is duplicated 3 times and you are missing a comma to continue this line.

File: app/assets/stylesheets/hubstats/bootstrap.css [hunk 4]
```diff
@@ -654,7 +654,8 @@ blockquote .small:before {
   content: '\2014 \00A0';
 }
 .blockquote-reverse,
-blockquote.pull-right {
+blockquote.pull-right,
+blockquote.deploy-right {
```
Comment (bbergstrom) [inline app/assets/stylesheets/hubstats/bootstrap.css]: Are these actually needed since we determined here that `pull` is a bootstrap class and not in reference to a "pull request"?

File: app/assets/javascripts/hubstats/deploys.js (hunk 1 above)
Comment (emmasax1) [inline app/assets/javascripts/hubstats/deploys.js]: I wasn't sure if they would work differently for the different group_bys on the pull requests page versus the deploys page or not. But if we're sure that they'll transfer over and work for the deploys page as well, then I can either comment them out or delete them.

File: app/controllers/hubstats/deploys_controller.rb [hunk 5]
```diff
@@ -4,22 +4,22 @@ module Hubstats
   class DeploysController < ApplicationController
 
     def index
-      #deploy_id = Hubstats::Deploy
-      #  .order_with_timespan(@timespan, "ASC")
-      #  .belonging_to_users(params[:users])
-      #  .belonging_to_repos(params[:repo])
-      #  .has_many_pull_requests(params[:pull_requests])
-      #  .map(&:id)
+      deploy_id = Hubstats::Deploy
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: This appears to be un-used.

File: app/controllers/hubstats/deploys_controller.rb [hunk 6]
```diff
@@ -27,6 +27,14 @@ def index
     end
 
     def show
+      @deploy = Hubstats::Deploy.find(params[:id])
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: You can reduce the number of trips to the database here by including the associations you know you want.

Something like `Hubstats::Deploy.find(params[:id]).includes(:repo).includes(:pull_requests)` should work.

File: app/controllers/hubstats/deploys_controller.rb [hunk 7]
```diff
@@ -27,6 +27,14 @@ def index
...
+      @repo = @deploy.repo
+      @pull_requests = Hubstats::PullRequest.belonging_to_deploy(@deploy.id)
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: Use your association and write less code: `@pull_requests = @deploy.pull_requests`

File: app/controllers/hubstats/deploys_controller.rb (hunk 5 above)
Comment (emmasax1) [inline app/controllers/hubstats/deploys_controller.rb]: I actually thought it was, but you're right, the index isn't relying on it at all.

File: app/controllers/hubstats/deploys_controller.rb [hunk 8]
```diff
@@ -27,6 +27,14 @@ def index
...
+      @pull_request_count = Hubstats::PullRequest.belonging_to_deploy(@deploy.id)
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: Don't go back to the database when you already have the object you need data from.  Also, do you mean to call size on this? `pull_request_count = @pull_requests.size`

File: app/controllers/hubstats/deploys_controller.rb [hunk 9]
```diff
@@ -27,6 +27,14 @@ def index
...
+      @stats = {
+        pull_request_count: @pull_request_count
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: If you don't need pull_request_count in your view, don't make it an instance variable.

File: app/controllers/hubstats/deploys_controller.rb (hunk 6 above)
Comment (emmasax1) [inline app/controllers/hubstats/deploys_controller.rb]: Then would I just use the params thing, or would I still have to go back to @deploy?

Comment (emmasax1) [inline app/controllers/hubstats/deploys_controller.rb]: Never mind, you answered the question below.

File: app/models/hubstats/pull_request.rb [hunk 10]
```diff
@@ -23,7 +24,8 @@ class PullRequest < ActiveRecord::Base
 
     belongs_to :user
     belongs_to :repo
-    has_and_belongs_to_many :labels, :join_table => 'hubstats_labels_pull_requests'
+    belongs_to :deploy
+    has_and_belongs_to_many :labels, :join_table => "hubstats_labels_pull_requests"
```
Comment (bbergstrom) [inline app/models/hubstats/pull_request.rb]: Does this work?  Deploys don't have labels themselves, but the pull requests within the deploy have labels.

File: config/routes.rb [hunk 11]
```diff
@@ -1,7 +1,8 @@
 Hubstats::Engine.routes.draw do
   root to: "pull_requests#index" #sets default root to be the pulls page
   post "/handler" => "events#handler", :as => :handler
-  resources :deploys, :only => [:create, :index] #routes to both index and to create method
+  resources :deploys, :only => [:create, :index, :show] #routes to both index and to create method
+  #get "/:id" => "deploys#show", :as => :deploy #routes to the specific deploy
```
Comment (bbergstrom) [inline config/routes.rb]: Delete.

File: app/controllers/hubstats/deploys_controller.rb (hunk 6 above)
Comment (emmasax1) [inline app/controllers/hubstats/deploys_controller.rb]: It isn't liking those changes. It says that includes is undefined.

File: spec/controllers/hubstats/deploys_controller_spec.rb [hunk 12]
```diff
@@ -85,11 +111,29 @@ module Hubstats
                                   :deployed_at => "2011-02-03 03:00:00 -0500",
                                   :deployed_by => "emmasax1")
         deploys_ordered = [deploy2, deploy3, deploy1]
-        expect(Hubstats::Deploy).to receive(:order_with_timespan).and_return(deploys_ordered)
+        # failing as a result of adding in grouping by repo
```
Comment (bbergstrom) [inline spec/controllers/hubstats/deploys_controller_spec.rb]: Let's fix this.

File: spec/lib/hubstats/github_api_spec.rb [hunk 13]
```diff
@@ -59,10 +59,10 @@ module Hubstats
           allow(ENV).to receive(:[]).and_return(nil)
         end
 
-        it 'should fail to initialize at all' do
-          Hubstats::GithubAPI.configure()
-          expect{Hubstats::GithubAPI.client}.to raise_error Octokit::Unauthorized
-        end
+        #it 'should fail to initialize at all' do
```
Comment (bbergstrom) [inline spec/lib/hubstats/github_api_spec.rb]: Leave a comment here about why we commented it out.

File: app/assets/javascripts/hubstats/deploys.js (hunk 1 above)
Comment (bbergstrom) [inline app/assets/javascripts/hubstats/deploys.js]: Assuming the HTML markup structure and the parameters are the same it should work for both.  Since it is now going to be used by Pull Requests and Deploys, it would make sense to move it to a shared file like application.js.

File: app/models/hubstats/pull_request.rb (hunk 10 above)
Comment (emmasax1) [inline app/models/hubstats/pull_request.rb]: Yes. The label is still attached to the pull request, so saying that a pull request also belongs to a deploy doesn't affect the labels.

File: app/controllers/hubstats/deploys_controller.rb (hunk 6 above)
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: Maybe I have it backwards, see the index action above for an example.

File: app/models/hubstats/pull_request.rb (hunk 10 above)
Comment (bbergstrom) [inline app/models/hubstats/pull_request.rb]: OH, this is the PullRequest model.  Ignore me. :)

File: app/assets/stylesheets/hubstats/bootstrap.css [hunk 14]
```diff
@@ -654,7 +654,8 @@ blockquote .small:before {
   content: '\2014 \00A0';
```
Comment (elliothursh) [inline app/assets/stylesheets/hubstats/bootstrap.css]: I think it's best practice not to modify the boostrap.css file. That way when bootstrap bumps the version we can just replace this file with the new version.

File: app/assets/javascripts/hubstats/deploys.js (hunk 1 above)
Comment (emmasax1) [inline app/assets/javascripts/hubstats/deploys.js]: I don't think we need to move it. The deploys page seems to continue to work with the code commented out, so it might not need it to be in application.js at all.

File: app/assets/stylesheets/hubstats/bootstrap.css (hunk 14 above)
Comment (emmasax1) [inline app/assets/stylesheets/hubstats/bootstrap.css]: There are a couple changes here that actually do need a deploy thingy so that the deploy page works the same as the pull request page.

File: app/assets/javascripts/hubstats/deploys.js (hunk 1 above)
Comment (bbergstrom) [inline app/assets/javascripts/hubstats/deploys.js]: Technically, it doesn't need to be.  It could be in any file.  But its confusing for developers to have code in a file called 'pull_requests' that is also used else where.  For the sake of human readability and organization, we should move the shared code to a file that isn't dedicated to one resource/controller by its name.

File: app/assets/stylesheets/hubstats/bootstrap.css [hunk 15]
```diff
@@ -5641,7 +5641,8 @@ button.close {
   margin-right: auto;
   margin-left: auto;
 }
-.pull-right {
+.pull-right,
```
Comment (bbergstrom) [inline app/assets/stylesheets/hubstats/bootstrap.css]: Remove this last change to the bootstrap stylesheet.

File: app/controllers/hubstats/deploys_controller.rb [hunk 16]
```diff
@@ -5,21 +5,22 @@ class DeploysController < ApplicationController
 
     def index
       #deploy_id = Hubstats::Deploy
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: Delete this commented out code unless you have a reason to keep it in the project.

File: app/controllers/hubstats/deploys_controller.rb [hunk 17]
```diff
@@ -5,21 +5,22 @@ class DeploysController < ApplicationController
...
-      #  .order_with_timespan(@timespan, "ASC")
-      #  .belonging_to_users(params[:users])
       #  .belonging_to_repos(params[:repo])
-      #  .has_many_pull_requests(params[:pull_requests])
+      #  .belonging_to_users(params[:users])
+      #  .order_with_timespan(@timespan, "ASC")
       #  .map(&:id)
 
       # sets to include user and repo, and sorts data
-      @deploys = Hubstats::Deploy.includes(:repo)
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: Why did you remove the eager loading of the associated repos?

File: app/controllers/hubstats/deploys_controller.rb [hunk 18]
```diff
@@ -27,6 +28,41 @@ def index
     end
 
     def show
+      @deploy = Hubstats::Deploy.find(params[:id])
+      @repo = @deploy.repo
+      @pull_requests = @deploy.pull_requests
+      @pull_request_count = @pull_requests.length
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: These 3 instance variables should be locally scoped.  They are put in the stats instance variable to make it available to your view.  In general, you shouldn't make extra instance variables in controller actions as they cause more memory churn on each request.

File: spec/controllers/hubstats/deploys_controller_spec.rb [hunk 19]
```diff
@@ -85,11 +111,31 @@ module Hubstats
...
+
+      it "should show the pull requests of the specific deploy" do
+        deploy = create(:deploy, :git_revision => "c1a2b37",
+                        :repo_id => 101010,
+                        :deployed_at => "2009-02-03 03:00:00 -0500",
+                        :deployed_by => "emmasax1")
+
+        # These fail because we're not calling the actual table in the controller; the factory girl for pull request doesn't have a deploy_id.
+        # However, it would work if a pull request did belong to a deploy
+        #get :show, id: deploy
```
Comment (bbergstrom) [inline spec/controllers/hubstats/deploys_controller_spec.rb]: Does this need to be `get :show, id: deploy.id`?

File: app/controllers/hubstats/deploys_controller.rb [hunk 20]
```diff
@@ -27,6 +22,41 @@ def index
...
+      @net_additions = find_net_additions(@deploy.id)
+      @comment_count = find_comment_count(@deploy.id)
+      @stats = {
+        pull_count: @pull_request_count,
+        net_additions: @net_additions,
+        comment_count: @comment_count
+      }
+    end
+
+    def find_net_additions(deploy_id)
```
Comment (bbergstrom) [inline app/controllers/hubstats/deploys_controller.rb]: Use local not instance scoped variables in this method and `find_comment_count`.  You are just using them for their return value.

File: spec/controllers/hubstats/deploys_controller_spec.rb (hunk 19 above)
Comment (emmasax1) [inline spec/controllers/hubstats/deploys_controller_spec.rb]: No. I still get an ``ActiveRecord::StatementInvalid: Mysql2::Error: Unknown column 'hubstats_pull_requests.deploy_id' in 'where clause': SELECT `hubstats_pull_requests`.* FROM `hubstats_pull_requests`  WHERE `hubstats_pull_requests`.`deploy_id` = 846`` error.

Comment (bbergstrom) [inline spec/controllers/hubstats/deploys_controller_spec.rb]: Did you migrate your test database? I think we did this with `rake db:test:prepare`?

File: app/controllers/hubstats/deploys_controller.rb (hunk 18 above)
Comment (emmasax1) [inline app/controllers/hubstats/deploys_controller.rb]: A few actually do need to be instance variables so that the show.html.erb can see them. Otherwise the page breaks.

File: spec/controllers/hubstats/deploys_controller_spec.rb (hunk 19 above)
Comment (emmasax1) [inline spec/controllers/hubstats/deploys_controller_spec.rb]: Oh yes. This is correct. I forgot we added the new migration.

**Comment by emmasax1** [comment]
All righty. I think that should be good! Can you do a final CR/QA @bbergstrom?

**Comment by bbergstrom** [comment]
CRQA :+1:
